BIP drafts: SwiftSync Specification (#2152)
bitcoin/bips: bip-0456.md

```
  BIP: ?
  Layer: Peer Services
  Title: Hints for unspent coins
  Authors: Robert Netzke <rob@2140.dev>, Ruben Somsen <ruben@2140.dev>
  Deputies: Edil Medeiros <edil@vinteum.org>
  Status: Draft
  Type: Specification
  Assigned: ?
  License: BSD-3-Clause
  Discussion: https://groups.google.com/g/bitcoindev/c/FpSWUxItXQs/m/pnfjP6rFCgAJ
```

## Abstract

The SwiftSync protocol requires a client to have foresight, or "hints", into the UTXO set at a state at a particular
height, which is verified at the end of the protocol. This document describes a concise representation of the UTXO set.
Clients performing SwiftSync may use this file of hints to perform IBD and verify the UTXO set they arrive at is
correct.

## Motivation

SwiftSync can improve the user experience by accelerating IBD, however the protocol requires the client verify a UTXO
set corresponds to the blockchain history they received. Rather than simply encoding and distributing the UTXO set, a
far smaller representation may be computed and shared. Intuitively, just as how it is cheap for programs to share

> pointers to objects in memory, we define a "hintsfile" that encodes pointers to unspent outputs.

Reviewer: The Abstract and Motivation feel a bit mixed up here. Since you’ll have to fill out three Abstracts and three motivations, it might be the easiest to decide which BIP is the first one to be read, and put the motivation of the overarching proposal there.

Then this one could, e.g., simply read: "This document specifies the hintsfile for Swiftsync. The hintsfile stores a compact representation of the UTXO set at a particular height by encoding which TXOs from each block remain unspent."

On the other hand, the motivation should be more focused on what problem this proposal addresses and why this should be adopted, i.e., that the hints file enables skipping all intermediate UTXO set states per foreseeing whether a UTXO is to be retained, that this speeds up IBD, etc. should be in the motivation.

## Specification

The key words "MUST", "MUST NOT", "REQUIRED", "SHALL", "SHALL NOT", "SHOULD", "SHOULD NOT", "RECOMMENDED", "MAY", and
"OPTIONAL" in this document are to be interpreted as described in RFC 2119.

Our goal is to indicate which outputs in a block, or set of blocks, will remain unspent at a particular chain height.
There are many ways to do so, however an intuitive representation is a
[bitset](https://en.cppreference.com/w/cpp/utility/bitset.html). Each output in a block is assigned a bit, with a `0`
bit denoting the output will be spent, and a `1` to denote the output is unspent. For a block of 8 outputs, we may

> arrive at a bitset of `1000 0010`, which conveys the 0th and 6th index are UTXOs.

Reviewer: I looked up the outputs per day for the last ~1000 days from mainnet.observer and divided them by 144 for an educated guess. Some of those days might have more or fewer blocks, so this is probably fairly inaccurate, but by that method I found a minimum of 7200 and a maximum of 10,700 TXOs created per block. I think giving this number would help readers’ intuition why encoding it this way would be very inefficient. I [calculated](https://bitcoin.stackexchange.com/a/118452/5406) that we could create slightly over 32,000 TXOs in a single block in the past. Maybe consider adding more context along the lines of:

```suggestion
arrive at a bitset of `1000 0010`, which conveys the 0th and 6th index are UTXOs. Actual blocks recently have about 8000–10,000 outputs, but the theoretical maximum is over 32k. Especially for older blocks, very few of these TXOs would remain unspent, so a bitset encoding would take several kilobits of mostly 0s.
```

You could also explain this alternative approach in the Rationale, or take it to the Motivation to explain why Elias-Fano was used. In the specification, it seems  a bit out of place.

Empirically, this representation does poorly, as the vast majority of outputs are spent. In fact, simply recording the
indices within the block that will be unspent is a more optimal approach than the bitset. For our example above, that
would be recording the numbers `0, 6`. To efficiently represent this data, we may formulate the problem as the
following: what is the most efficient way to represent $n$ random, monotonically increasing indices less than or equal
to size $m$? In other words, for a block with $n$ unspent outputs, what is an efficient representation?

This formulation has a known theoretical optimum of $\\log_2 \\binom{m}{n}$ bits. _Elias-Fano_ developed a
representation that is reasonably close to this optimum in a concise and maintainable format.

> ### Elias-Fano Encoding

Reviewer: I found this description of the encoding too dense to follow. I had to work the example by hand to understand the instructions. Also, some of the instructions are missing and are only mentioned in the example. More detailed suggestions inline.

Suppose we would like to encode $n$ monotonically increasing elements ${ i\_{0}, ..., i\_{n - 1} }$ where $i\_{n - 1} =

> m$. _Elias-Fano_ builds on the intuition that numbers may share bits in their binary representation. Consider the

Reviewer: ```suggestion
m$. _Elias-Fano_ builds on the intuition that in a sequence of monotonically increasing numbers the least significant bits differ while the most significant bits of numbers increase only gradually. Consider the
```

integers 6 and 7. The binary representation of 6 is `0110` while the binary representation of 7 is `0111`, differing by

> only one bit. This observation may be used to efficiently store elements. For the lowest bits of the elements, we will

Reviewer: This example didn’t help me. I would suggest either dropping it or elaborating to explain the distinction between the most significant bits and less significant bits.

```suggestion
m$. _Elias-Fano_ builds on the intuition that numbers may share bits in their binary representation. This observation may be used to efficiently store elements. For the lowest bits of the elements, we will
```

```suggestion
m$. _Elias-Fano_ builds on the intuition that numbers may share bits in their binary representation. Consider for example the three random
integers 2 , 5, and 7. In binary of 2 is `0010`, 5 is `0101`, and 7 is `0111`. Splitting them into the two most significant bits and two least significant bits, the most significant two bits are slowly counting up `00`, `01`, and `01`, while the least significant two bits don’t follow a pattern, `10`, `01`, and `11`. This observation may be used to efficiently store elements. For the lowest bits of the elements, we will
```

> represent them in a bitset, and for the highest bits, we will represent the difference in bits as unary.

Reviewer: Just their difference?

```suggestion
represent them in a bitset, and for the highest bits, we will represent their difference as unary.
```

We define the following:

- Let ${ i\_{0}, ..., i\_{n-1} }$ be a list of $n$ elements where $i\_{n-1} = m$
- Let $\\ell(m, n)$ be the function that determines the number of low bits to use: $\\ell(m, n) = \\left\\lfloor \\log_2
  \\ \\left(\\frac{m+1}{n}\\right) \\right\\rfloor$

> - Let $\\text{unary}(q)$ be the following function: $\\text{unary}(q) = \\underbrace{0\\cdots0}\_{q}1$

Reviewer: This renders oddly: 

<img width="825" height="136" alt="Image" src="https://github.com/user-attachments/assets/8081bcef-1a01-4586-8229-69f0f354522b" />

It looks like the 1 and q are not related at all.

Reviewer: This section seems to have more than the necessary escaping. I think this should work fine:
```suggestion
- Let ${ i_{0}, ..., i_{n-1} }$ be a list of $n$ elements where $i_{n-1} = m$
- Let $\ell(m, n)$ be the function that determines the number of low bits to use: $\ell(m, n) = \left\lfloor \log_2 \left(\frac{m+1}{n}\right) \right\rfloor$
- Let $\text{unary}(q)$ be the following function: $\text{unary}(q) = \underbrace{0\cdots0}_{q}1$
```

Reviewer 2: This one is rendering fine to me, but I'll try to fiddle it to be more stable.
<img width="437" height="48" alt="Screenshot 2026-07-27 at 13 29 34" src="https://github.com/user-attachments/assets/762d763f-2d21-42c9-a5a7-d23e1417d843" />

Reviewer 2: Addressed that by getting rid of the `unary` function altogether. I think it's now easier to grasp the whole process.

> We start by determining $\\ell$ for a given $n$ and $m$. For each $i$, take the $\\ell$ least significant bits and

Reviewer: Maybe this is obvious to smarter people, but it might be helpful to keep the reader abreast of how many bits are being encoded. ;)

```suggestion
Numbers up to $m$ are represented with up to $\\lceil log_2(m+1)\\rceil$ bits in binary.
We start by determining how the bits are split by calculating $\\ell$ for a given $n$ and $m$. For each $i$ in ascending order, take the $\\ell$ least significant bits and
```

append them, most significant bit ordering, to a bitset $L$. Next, iterate over each element and take the remaining most

> significant bits, then record the difference between the last element's high bits and the current element's high bits.

Reviewer: ```suggestion
append them, most significant bit ordering, to a bitset $L$. Next, iterate over each element and interpret the remaining most
significant bits as a number, then record the difference between the last element's high bits and the current element's high bits. The first element records the difference to 0 as there is no preceding element.
```

Record that number using the $\\text unary$ function, and append the unary to a vector $H$. The final representation is
$m$, $n$, $L$, $H$.

### Worked Example

Suppose $S = [3, 7, 12]$. Then $n=3$ and $m=12$.

#### 1. Compute the number of lower bits to use

We start by computing $\\ell$:

$\\ell = \\left\\lfloor \\log_2\\left(\\frac{m+1}{n}\\right) \\right\\rfloor = \\left\\lfloor
\\log_2\\left(\\frac{13}{3}\\right) \\right\\rfloor = \\lfloor \\log_2(4.33) \\rfloor = \\lfloor 2.11 \\rfloor = 2$

> 

Reviewer: ```suggestion

We write out the binary representation for all elements in the sequence $S$ and split them into upper and lower bits:

```

## Distribution

A malicious hintsfile distributor cannot lead a client to accepting an invalid UTXO set state, but they can expend time
and resources of a client attempting to perform IBD with SwiftSync. Clients SHOULD obtain a hintsfile from a party that
has a moral, financial, or otherwise incentive-aligned reason to provide truthful data.

## Reference Implementation and Test Vectors

### Reference Implementation(s)

- [Bitcoin Core](https://github.com/rustaceanrob/bitcoin/tree/hintsfile-v1)
- [`hintsfile crate`](https://github.com/rustaceanrob/hintsfile/tree/master)

### Test Vectors

- [Elias-Fano encoding](test_vectors/hinstfile/elias_fano.json)

## Acknowledgements

Thank you to l0rinc for challenging and reviewing the many iterations of this file, and thank you to Eliam for exploring
alternatives and giving insights.

## References

- [Research repository](https://github.com/rustaceanrob/swiftsync-research)
- [Explanatory article](https://t.holmium.no/dia/elias-fano/#_representation_1_fixed_size_compact_encoding)
- [Paper](https://drops.dagstuhl.de/storage/00lipics/lipics-vol078-cpm2017/LIPIcs.CPM.2017.30/LIPIcs.CPM.2017.30.pdf)

> 

Reviewer: Please include a Compatibility Section, even if just to say that there are no compatibility issues.
